### ISSUE TITLE
Add warning for variables set / changed in a function which are not declared as local.

### ISSUE THREAD

**Issue by dougborg** [koalaman/shellcheck#213]
This is similar to issue #115, but my take is slightly different. I feel that any variable modified inside a function should trigger a ShellCheck warning if it is not declared `local`. This includes explicit modifications using `=`, `let`, and `for` loops.

Here is an example that illustrates why I think this is a problem:

``` bash
#!/bin/bash

MY_SERVERS='server1 server2 server3'

trust_my_server() {
  my_server="${1}"
  shift
  servers="${@}"

  for server in ${servers}; do
    ssh ${server} 'trust ${my_server}'
  done
}

do_stuff() {
  for server in $MY_SERVERS; do
    trust_my_server ${server} ${MY_SERVERS//${server}/}
    do_something_else ${server} # This will not do what you expect; ${server} has been changed on you!
  done
} 

do_stuff
```

In the example above, the `do_something_else` function does not get called with the element of `$MY_SERVERS` you expect it to from the main `for` loop. As side effect of running `trust_my_server` (because server is not declared `local`), `server` will get set to the last element of what was passed to `trust_my_server` by its `for` loop by the time `do_something_else` is called, leading to much woe.
